File: 2026/08-August/05-Wednesday/solution.kt
class Solution {
    fun remainingMethods(
        n: Int,
        k: Int,
        invocations: Array<IntArray>
    ): List<Int> {

        val graph = Array(n) { mutableListOf<Int>() }
        val undirected = Array(n) { mutableListOf<Int>() }

        for (edge in invocations) {
            val u = edge[0]
            val v = edge[1]

            graph[u].add(v)
            undirected[u].add(v)
            undirected[v].add(u)
        }

        val suspicious = BooleanArray(n)

        fun dfs(u: Int) {
            suspicious[u] = true
            for (v in graph[u]) {
                if (!suspicious[v]) {
                    dfs(v)
                }
            }
        }

        dfs(k)

        val visited = BooleanArray(n)

        fun dfs2(u: Int) {
            visited[u] = true
            for (v in undirected[u]) {
                if (!visited[v]) {
                    suspicious[v] = false
                    dfs2(v)
                }
            }
        }

        for (i in 0 until n) {
            if (!suspicious[i] && !visited[i]) {
                dfs2(i)
            }
        }

        val ans = mutableListOf<Int>()
        for (i in 0 until n) {
            if (!suspicious[i]) {
                ans.add(i)
            }
        }

        return ans
    }
}

File: 2026/08-August/05-Wednesday/solution.cpp
class Solution {
public:
    vector<int> remainingMethods(int n, int k, vector<vector<int>>& invocations) {
        vector<vector<int>> graph(n), undirected(n);

        for (auto& e : invocations) {
            int u = e[0], v = e[1];
            graph[u].push_back(v);
            undirected[u].push_back(v);
            undirected[v].push_back(u);
        }

        vector<bool> suspicious(n, false), visited(n, false);

        auto dfs = [&](auto&& self, int u) -> void {
            suspicious[u] = true;
            for (int v : graph[u]) {
                if (!suspicious[v]) {
                    self(self, v);
                }
            }
        };

        dfs(dfs, k);

        auto dfs2 = [&](auto&& self, int u) -> void {
            visited[u] = true;
            for (int v : undirected[u]) {
                if (!visited[v]) {
                    suspicious[v] = false;
                    self(self, v);
                }
            }
        };

        for (int i = 0; i < n; i++) {
            if (!suspicious[i] && !visited[i]) {
                dfs2(dfs2, i);
            }
        }

        vector<int> ans;
        for (int i = 0; i < n; i++) {
            if (!suspicious[i]) {
                ans.push_back(i);
            }
        }

        return ans;
    }
};
